### DISCUSSION TITLE
Lesson12 - ERC20Mock (General Questions)

### DISCUSSION THREAD

**Discussion by Glow-in-the-dark** [Cyfrin/foundry-full-course-cu#314] [Q&A] [answered]
I realised the ERC20Mock which was used in the lesson was similar to this:
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/0a25c1940ca220686588c4af3ec526f725fe2582/contracts/mocks/ERC20Mock.sol
Therefore for the constructor, we initiate/inputs 4 different arguments with it. 

however, when i run:
> forge install openzeppelin/openzeppelin-contracts --no-commit

my ERC20Mock.sol looks like this:
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/mocks/token/ERC20Mock.sol

1) Did OpenZeppelin update it, and therefore the latest one i got was the updated one ?
2) Why did they update it this way ? The older one seems more flexible to use isn't it ?
3) I suppose with this new ERC20Mock:
```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

import {ERC20} from "../../token/ERC20/ERC20.sol";

contract ERC20Mock is ERC20 {
    constructor() ERC20("ERC20Mock", "E20M") {}

    function mint(address account, uint256 amount) external {
        _mint(account, amount);
    }

    function burn(address account, uint256 amount) external {
        _burn(account, amount);
    }
}
```

It seems like it's hardcoded, and we can't use it for mocking our weth / wbtc contracts.
I don't get why they want to set a Mock explicitly tho. I mean i suppose we can always change the code to suit us. But just wondering why the author made the change, when i seems more elegant for the previous older version contract. Afterall when u want to deploy a ERC20 contract, you would definitely need to input the Amount & the contract deployer address anyway.

4) If this is the new and updated one, how would i then ensure the next time when i run:
> forge install openzeppelin/openzeppelin-contracts --no-commit

i get the old version instead? 

Thanks.

**Comment by Glow-in-the-dark** [comment]
Also, i realised the older contract which the contructor has a payable.
```solidity
// mock class using ERC20
contract ERC20Mock is ERC20 {
    constructor(
        string memory name,
        string memory symbol,
        address initialAccount,
        uint256 initialBalance
    ) payable ERC20(name, symbol) {
        _mint(initialAccount, initialBalance);
    }
    ```
    
    why do they need a payable there? Noone should sending tokens to the token contract isn't it ?

**Reply by jaimebarrancos** [reply]
I have the same question, what is weird for me is the openzeppelin github has the version with the 4 arguments (https://github.com/OpenZeppelin/openzeppelin-contracts/blob/0a25c1940ca220686588c4af3ec526f725fe2582/contracts/mocks/ERC20Mock.sol) 
For now I will manually copy the mock used in the course.

**Reply by shikhar229169** [reply]
Yes, I also did the same thing.

**Comment by 0xusmanf** [comment, marked as answer]
This issue should be updated in the [chronological-updates.md](https://github.com/Cyfrin/foundry-full-course-f23/blob/main/chronological-updates.md), as everybody watching the tutorial after the change will be confused.
In the latest version of openzeppelin/openzeppelin-contracts `ERC20Mock.sol` file has been updated with **0** constructor parameters.
You should use `forge install openzeppelin/openzeppelin-contracts@v4.8.3 --no-commit` command to download the appropriate version synced with the video. @PatrickAlphaC

**Reply by cromewar** [reply]
Thanks for the answer @usmanfarooq91

**Reply by thurendous** [reply]
This is helpful! Thanks
